### ISSUE TITLE
Generate transform_to_physical

### ISSUE THREAD

**Issue by kynan** [gmarkall/manycore_form_compiler#14]
Currently, quadrature (points and weights) and shape (function coefficients and their derivatives) information needed for the transformation from reference to physical space (`transform_to_physical`) is pulled out from Fluidity state at runtime. That is inefficient, since this information is static for a given element and hence available at code generation time.

`transform_to_physical` (which is currently a library function) needs to have this data as literal arrays of doubles inserted by the form compiler and therefore needs to be a generated function specific to a particular element. Transformation from reference to physical space should be done in the following way:
- `transform_to_physical` needs to be generated:
  - with `weight`, `l`, `n` and `dn` as arrays of literals
  - for the right topological dimension (eliminate `if` switches)
  - computing shape function derivatives only if required
- if there are forms with different elements, a separate `transform_to_physical` is needed for each unique element
- The assembler needs to make sure that before a kernel is called, the _right_ `transform_to_physical` has been called with the _right_ parameters
## Signature and parameters
- computing only _det(J)_, the signature is
  
  ``` c++
  __global__ void transform_to_physical(
                double *node_coords,
                double *detJ,
                int n_ele);
  ```
- computing _det(J)_ and _inv(J)_, the signature is
  
  ``` c++
  __global__ void transform_to_physical(
                double *node_coords,
                double *detJ,
                double *invJ,
                int n_ele);
  ```

The parameters are
- `node_coords` coordinate field (input, expanded from nodes to nodes in elements, _#dims x #nodes_per_ele x #ele_)
- `detJ` Jacobian determinant for each element (output, _#ele_)
- `invJ` Jacobian inverse for each element (output, _#dims x #dimx x #ele_)
## Transform class

A class `Transform` should be added
- holding a `dict` with UFL element representation as keys and meta data as values:
  - `invJ` needed (i.e. does any form using that element have derivatives)?
- the element determines whether we need to compute J
  - only once per element (i.e. at 1st quadrature point), if the cell is a simplex (interval, triangle, tetrahedron) and the shape functions are linear
  - at every quadrature point otherwise

**Comment by kynan** [comment]
The assumption that a separate `transform_to_physical` is needed for each unique element is wrong. It is always ever only necessary to have one `transform_to_physical` kernel that depends on the element / function space of the coordinate field, which makes our lives much easier since it does not break our abstractions.

If there were potentially multiple transformation kernels as thought initially and it was necessary to figure out whether they need to compute derivatives or not and which one to call for which form etc. this would be very hard to fit to the UFL abstractions of form, element etc.

That means the `Transformation` class is the right abstraction for this.

**Comment by gmarkall** [comment]
This is done in the _generate-transforms_ branch (merged in e739ee4), which actually incorporates the generation of the transform in the local assembly kernel.
